File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        if not list2:
            return list1

        curr1 = list1
        curr2 = list2

        if curr1.val <= curr2.val:
            merged_head = ListNode(curr1.val)
            curr1 = curr1.next
            
        else: 
            merged_head = ListNode(curr2.val)
            curr2 = curr2.next
            
        merged_curr = merged_head
        merged_prev_curr = merged_head

        i = 0

        while curr1 or curr2:
            if not curr1:
                merged_curr = curr2
                curr2 = curr2.next
            elif not curr2:
                merged_curr = curr1
                curr1 = curr1.next

            elif curr1.val <= curr2.val:
                merged_curr = curr1
                curr1 = curr1.next

            else:
                merged_curr =curr2
                curr2 = curr2.next

            print(i,merged_curr.val)
            i+=1


            if merged_prev_curr:
                merged_prev_curr.next = merged_curr

            merged_prev_curr = merged_curr


        return merged_head

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* prevHead = new ListNode(0, head);
        ListNode* curr = prevHead;
        ListNode* fast = head;

        for(int i=0;i<n;i++){
            if(!fast) return head;
            fast = fast->next;
            //std::cout << fast->val <<"\n";
        }

        while(fast != nullptr){
            //std::cout << fast->val << ", "<< curr->val <<"\n";
            curr = curr->next;
            fast = fast->next;
        }

        curr->next = curr->next->next;

        return prevHead->next;
    }
};
